File: 백준/Silver/21736. 헌내기는 친구가 필요해/헌내기는 친구가 필요해.py
# 헌내기는 친구가 필요해
import sys
input = sys.stdin.readline
from collections import deque

n, m = map(int,input().split())
graph = []
visited = [[0] * m for _ in range(n)]
meets = 0 # 사람을 만난 횟수 

for i in range(n):
    row = input().rstrip()
    graph.append(row)
    for j in range(m):
        if row[j] == "I":
            doyeon = (i, j) # 도연이의 시작 위치

que = deque([doyeon])

while que:
    cr, cc = que.popleft()

    if visited[cr][cc]:
        continue
    else:
        visited[cr][cc] = 1

    if graph[cr][cc] == "P":
        meets += 1

    for x, y in [(1, 0), (-1, 0), (0, 1), (0, -1)]:
        nr, nc = cr + x, cc + y

        if (0 > nr or nr >= n or 0 > nc or nc >= m):
            continue
        elif visited[nr][nc]:
            continue
        elif graph[nr][nc] == "X":
            continue

        que.append((nr, nc))

if meets:
    print(meets)
else:
    print("TT")

File: 백준/Silver/21736. 헌내기는 친구가 필요해/헌내기는 친구가 필요해.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken()); // 행의 수
        int m = Integer.parseInt(st.nextToken()); // 열의 수

        char[][] graph = new char[n][m];
        boolean[][] visited = new boolean[n][m];
        int meets = 0; // 사람을 만난 횟수
        int startRow = 0, startCol = 0; // 도연이의 시작 위치

        for (int i = 0; i < n; i++) {
            String row = br.readLine();
            for (int j = 0; j < m; j++) {
                graph[i][j] = row.charAt(j);
                if (graph[i][j] == 'I') {
                    startRow = i;
                    startCol = j;
                }
            }
        }

        ArrayDeque<int[]> que = new ArrayDeque<>();
        que.add(new int[]{startRow, startCol});

        while (!que.isEmpty()) {
            int[] current = que.removeFirst();
            int cr = current[0];
            int cc = current[1];

            if (visited[cr][cc]) {
                continue;
            }
            visited[cr][cc] = true;

            // 사람을 만난 경우
            if (graph[cr][cc] == 'P') {
                meets++;
            }

            // 상하좌우 이동
            int[][] directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
            for (int[] dir : directions) {
                int nr = cr + dir[0];
                int nc = cc + dir[1];

                if (nr < 0 || nr >= n || nc < 0 || nc >= m) {
                    continue; // 범위를 벗어난 경우
                }
                if (visited[nr][nc] || graph[nr][nc] == 'X') {
                    continue; // 이미 방문했거나 장애물인 경우
                }

                que.addLast(new int[]{nr, nc});
            }
        }

        if (meets > 0) {
            System.out.println(meets);
        } else {
            System.out.println("TT");
        }

    }
}
